### PR TITLE
+htt #15917 Adding play json support

### PR DESCRIPTION
**Pull request by guillaumebreton**
- Add play json dependency
- Add play json support to marshal/unmarshal json
- Add test using JsonSupportSpec

Note : at line 40 in PlayJsonSupport.scala, I leave a "// FIXME: identify charset by instance, not by name!
" comment as I don't have any idea on how to fix this.

### CODE REVIEW

**Comment by ktoso** [comment]
OK TO TES

**Comment by ktoso** [comment]
OK TO TEST

File: akka-http-marshallers-scala/akka-http-play-json/src/main/scala/akka/http/scaladsl/marshallers/playjson/PlayJsonSupport.scala [hunk 1]
```diff
@@ -0,0 +1,50 @@
+/*
```
Comment (ktoso) [inline akka-http-marshallers-scala/akka-http-play-json/src/main/scala/akka/http/scaladsl/marshallers/playjson/PlayJsonSupport.scala]: Thanks, that's the right place to put it.

**Comment by guillaumebreton** [comment]
I checked the jenkins result, and I am not sure if I can do something about it.

File: project/AkkaBuild.scala [hunk 2]
```diff
@@ -425,9 +425,10 @@ object AkkaBuild extends Build {
     id = "akka-http-marshallers-scala-experimental",
     base = file("akka-http-marshallers-scala"),
     settings = parentSettings ++ Seq(
-      version := streamAndHttpVersion
+      version := streamAndHttpVersion,
+      resolvers += "Typesafe Repo" at "http://repo.typesafe.com/typesafe/releases/"
```
Comment (ktoso) [inline project/AkkaBuild.scala]: Are you sure we need to add the resolver here? I'd have thought we include it already hm..

File: project/AkkaBuild.scala [hunk 3]
```diff
@@ -1547,6 +1552,9 @@ object Dependencies {
     // For akka-http spray-json support
     val sprayJson   = "io.spray"                     %% "spray-json"                   % "1.3.1"       // ApacheV2
 
+    //For akka-http play-json support
+    val playJson    = "com.typesafe.play"            %% "play-json"                    % "2.3.8"       // ApacheV2
```
Comment (ktoso) [inline project/AkkaBuild.scala]: @jroper do you have an opinion about what `play-json` version should be supported here (it's for a simple "support" trait for akka-http). Play `2.4` is in RC1 mode AFAICS, and akka-http is experimental, also in RC mode - would it make sense to depdend on `2.4-RC1` here instead then?

// cc @rkuhn

**Comment by ktoso** [comment]
The build failure most likely was due to us switching to JDK8 for PR validation tonight - not related to your PR - don't worry, we'll sort it out :-)

File: akka-http-marshallers-scala/akka-http-play-json/src/main/scala/akka/http/scaladsl/marshallers/playjson/PlayJsonSupport.scala [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+import scala.language.implicitConversions
+import scala.concurrent.ExecutionContext
+import akka.stream.FlowMaterializer
+import akka.http.scaladsl.marshalling.{ ToEntityMarshaller, Marshaller }
+import akka.http.scaladsl.unmarshalling.{ FromEntityUnmarshaller, Unmarshaller }
+import akka.http.scaladsl.model.{ ContentTypes, HttpCharsets }
+import akka.http.scaladsl.model.MediaTypes.`application/json`
+import play.api.libs.json._
+import play.api.libs.functional.syntax._
```
Comment (pjan) [inline akka-http-marshallers-scala/akka-http-play-json/src/main/scala/akka/http/scaladsl/marshallers/playjson/PlayJsonSupport.scala]: As far as I can tell, I believe this import is not required here?

**Comment by jroper** [comment]
I'm not sure about this... it introduces a circular dependency between our builds.

While at a module level, there is no circular dependency, it means we can never have a build where Play depends on the latest released version of akka-http, and akka-http depends on the latest released version of play-json.

It may also introduce circular dependencies whenever dbuild is used against Akka and Play, and this will most likely break the Reactive Platform build.

It will also introduce chicken egg scenarios - when Scala 2.12 is released, you won't be able to upgrade to it in akka-http because you'll need a play-json 2.12 build, and Play won't be able to upgrade to it because we'll need an akka-http 2.12 build.

Either we need to pull play-json out of the Play git repository, or pull this out of the Akka git repository, so that they can have independent lifecycles, and break the circular dependency.

So, I am very against this change.

**Comment by rkuhn** [comment]
Since Play depends on Akka already in a high-level sense, I agree that this module should live in the Play ecosystem.

File: project/AkkaBuild.scala (hunk 2 above)
Comment (guillaumebreton) [inline project/AkkaBuild.scala]: I tried to add play-json dependency without it and it failed. I may have missed something, but it only appears in the atmos project.

**Comment by hepin1989** [comment]
yes,as a play plugin

**Comment by guillaumebreton** [comment]
@jroper @rkuhn  Thanks for the review. This is a perfectly valid point, sorry I missed it. Should I close the PR myself ?

@hepin1989 I'm not sure about the play plugin implementation. My point here was to support the play json lib directly in akka http, without the complete play stack. If I write a play plugin, I will be force to use it inside a play stack, am I right ?

**Comment by ktoso** [comment]
Good point @jroper, thanks for chiming in. Closing this PR for now and hoping it'll find a place to live :-)

**Comment by rkuhn** [comment]
I don’t think it should be a Play plugin, it should just live together with play-json itself so that it can without cycles depend upon akka-http. Play users should be able to use play-json already without such a plugin.

**Comment by hepin1989** [comment]
@guillaumebreton so I think you want to using play-json with akka-http right?then may you need a akka extension.

I don't know that play-json dependents on play.should that be ?play-ws seems bound to play.

**Comment by hepin1989** [comment]
a better way I think just like @jroper said.split out play-json then you could create an akka extension,which just dependent on play-json.
and play-json it's self shoudl not dependent on play but just jackson.

**Comment by ktoso** [comment]
An akka extension is not really needed here - the implicits in the trait must be brought into scope to be used, nothing an extension can help with. 

+1 for taking this to live next to play-json

**Comment by jroper** [comment]
play-json doesn't depend on Play, but is a part of the Play build out of convenience - as soon as you start splitting things out of builds, it adds friction, releases become more effort, sweeping changes become difficult to coordinate, versions and compatibility becomes harder to reconcile.  So you don't split things out unless you have a good reason to, and so far there's been no good reason to split play-json out of Play.

We could pull this into Play, to live alongside play-json, though it is a bit odd, because it has nothing to do with Play - Play's relation to akka-http is that it depends on lower level akka-http APIs, not these higher level ones.  So it would feel a little out of place.  If play-json was its own build, then it would be a completely different story.

My preference for now would be to have this live entirely as its own module, outside of akka-http and outside of Play.  If or when we pull play-json out of Play, then it would make sense to combine this into it.

**Comment by ktoso** [comment]
I was not aware of it being tied build-wise with Play. Thanks for explaining how things fit together :-)
Seems we have all info we need to make an informed decision about this project being best hosted as separate repo/project.

I hope you'll be able to pull it off @guillaumebreton! Let us know if you'd need any help or hints

**Comment by guillaumebreton** [comment]
Thanks a lot @jroper and @ktoso :) it makes things very clear ! I will extract it in a separate project :+1:

**Comment by guillaumebreton** [comment]
@ktoso  Found the issue #17106 which might be linked to this discussion.

**Comment by ktoso** [comment]
Linking to the repo: https://github.com/guillaumebreton/play-json-support

**Pull request closed without merging**